### PR TITLE
v0.12.x release of Jane Street packages, with support for OCaml 4.08

### PR DESCRIPTION
**Pull request by xclerc**
note 1: the previous versions are marked as incompatible with 4.08
note 2: includes https://github.com/ocaml/opam-repository/pull/14208 for the packages touched by this pull request

### CODE REVIEW

File: packages/core/core.v0.12.1/opam [hunk 1]
```diff
@@ -10,7 +10,7 @@ build: [
   ["dune" "build" "-p" name "-j" jobs]
 ]
 depends: [
-  "ocaml"        {>= "4.07.0"}
+  "ocaml"        {>= "4.07.0" & "4.08.0"}
```
Comment (kit-ty-kate) [inline packages/core/core.v0.12.1/opam]:
```suggestion
  "ocaml"        {>= "4.07.0" & < "4.08.0"}
```
did you want to say < 4.08? (same for the other files)

Comment (xclerc) [reply packages/core/core.v0.12.1/opam]: Sure; thanks!

**Comment by kit-ty-kate** [comment]
Also the build was failing with:
```
#=== ERROR while compiling toplevel_expect_test.v0.12.0 =======================#
# context              2.0.4 | linux/x86_64 | ocaml-base-compiler.4.07.1 | pinned(https://ocaml.janestreet.com/ocaml-core/v0.12/files/toplevel_expect_test-v0.12.0.tar.gz)
# path                 ~/.opam/4.07/.opam-switch/build/toplevel_expect_test.v0.12.0
# command              ~/.opam/4.07/bin/dune build -p toplevel_expect_test -j 71
# exit-code            1
# env-file             ~/.opam/log/toplevel_expect_test-18-8d4d99.env
# output-file          ~/.opam/log/toplevel_expect_test-18-8d4d99.out
### output ###
#       ocamlc src/.toplevel_expect_test.objs/byte/toplevel_expect_test__Main.{cmo,cmt} (exit 2)
# (cd _build/default && /home/opam/.opam/4.07/bin/ocamlc.opt -w -40 -g -bin-annot -I src/.toplevel_expect_test.objs/byte -I /home/opam/.opam/4.07/lib/base -I /home/opam/.opam/4.07/lib/base/caml -I /home/opam/.opam/4.07/lib/base/md5 -I /home/opam/.opam/4.07/lib/base/shadow_stdlib -I /home/opam/.opam/4.07/lib/base_bigstring -I /home/opam/.opam/4.07/lib/base_quickcheck -I /home/opam/.opam/4.07/lib/bin_prot -I /home/opam/.opam/4.07/lib/bin_prot/shape -I /home/opam/.opam/4.07/lib/core_kernel -I /home/opam/.opam/4.07/lib/core_kernel/base_for_tests -I /home/opam/.opam/4.07/lib/fieldslib -I /home/opam/.opam/4.07/lib/findlib -I /home/opam/.opam/4.07/lib/jane-street-headers -I /home/opam/.opam/4.07/lib/mlt_parser -I /home/opam/.opam/4.07/lib/ocaml-compiler-libs/common -I /home/opam/.opam/4.07/lib/ocaml-compiler-libs/shadow -I /home/opam/.opam/4.07/lib/ocaml-migrate-parsetree -I /home/opam/.opam/4.07/lib/ocaml/compiler-libs -I /home/opam/.opam/4.07/lib/ocaml/threads -I /home/opam/.opam/4.07/lib/parsexp -I /home/opam/.opam/4.07/lib/ppx_assert/runtime-lib -I /home/opam/.opam/4.07/lib/ppx_bench/runtime-lib -I /home/opam/.opam/4.07/lib/ppx_compare/runtime-lib -I /home/opam/.opam/4.07/lib/ppx_derivers -I /home/opam/.opam/4.07/lib/ppx_enumerate/runtime-lib -I /home/opam/.opam/4.07/lib/ppx_expect/collector -I /home/opam/.opam/4.07/lib/ppx_expect/common -I /home/opam/.opam/4.07/lib/ppx_expect/config -I /home/opam/.opam/4.07/lib/ppx_expect/matcher -I /home/opam/.opam/4.07/lib/ppx_expect/payload -I /home/opam/.opam/4.07/lib/ppx_hash/runtime-lib -I /home/opam/.opam/4.07/lib/ppx_here/expander -I /home/opam/.opam/4.07/lib/ppx_inline_test/config -I /home/opam/.opam/4.07/lib/ppx_inline_test/runner/lib -I /home/opam/.opam/4.07/lib/ppx_inline_test/runtime-lib -I /home/opam/.opam/4.07/lib/ppx_module_timer/runtime -I /home/opam/.opam/4.07/lib/ppx_sexp_conv/runtime-lib -I /home/opam/.opam/4.07/lib/ppxlib -I /home/opam/.opam/4.07/lib/ppxlib/ast -I /home/opam/.opam/4.07/lib/ppxlib/print_diff -I /home/opam/.opam/4.07/lib/ppxlib/traverse_builtins -I /home/opam/.opam/4.07/lib/re -I /home/opam/.opam/4.07/lib/result -I /home/opam/.opam/4.07/lib/seq -I /home/opam/.opam/4.07/lib/sexplib -I /home/opam/.opam/4.07/lib/sexplib0 -I /home/opam/.opam/4.07/lib/splittable_random -I /home/opam/.opam/4.07/lib/stdio -I /home/opam/.opam/4.07/lib/time_now -I /home/opam/.opam/4.07/lib/typerep -I /home/opam/.opam/4.07/lib/variantslib -I types/.toplevel_expect_test_types.objs/byte -I types/.toplevel_expect_test_types.objs/native -intf-suffix .ml -no-alias-deps -open Toplevel_expect_test -o src/.toplevel_expect_test.objs/byte/toplevel_expect_test__Main.cmo -c -impl src/main.pp.ml)
# File "src/main.ml", line 57, characters 34-37:
# Error: Unbound record field msg
```

File: packages/core/core.v0.12.1/opam (hunk 1 above)
Comment (kit-ty-kate) [reply packages/core/core.v0.12.1/opam]: could you do the same for the other packages?

Comment (xclerc) [reply packages/core/core.v0.12.1/opam]: Sorry, I misinterpreted your message;
https://github.com/ocaml/opam-repository/pull/14212/commits/5cac02cb7f008223397f623fb39f897d03b73079 should fix the constraints.

**Comment by xclerc** [comment]
Regarding `toplevel_expect_test`, I think the upper bound for
`ppxlib` was too lax. https://github.com/ocaml/opam-repository/pull/14212/commits/51eeef8306986c1fc23023617492efed8102c23c should fix the problem (at least
it does locally).

**Comment by xclerc** [comment]
(I don't think the failures reported by the CI are related to this PR.)

**Comment by XVilka** [comment]
```
+ /home/opam/.opam/4.07/bin/dune "build" "-p" "torch" "-j" "71" (CWD=/home/opam/.opam/4.07/.opam-switch/build/torch.0.1)
-   ocamlmklib src/wrapper/dlltorch_core_stubs.so,src/wrapper/libtorch_core_stubs.a (exit 2)
- (cd _build/default && /home/opam/.opam/4.07/bin/ocamlmklib.opt -g -o src/wrapper/torch_core_stubs src/wrapper/torch_stubs.o src/wrapper/torch_api.o -lstdc++ -Wl,-R/home/opam/.opam/4.07/lib/libtorch/lib -L/home/opam/.opam/4.07/lib/libtorch/lib -lc10 -lcaffe2 -ltorch)
- /usr/bin/ld: src/wrapper/torch_api.o: relocation R_X86_64_PC32 against symbol `_ZN3c10eqERKNS_12TensorTypeIdES2_' can not be used when making a shared object; recompile with -fPIC
- /usr/bin/ld: final link failed: Bad value
- collect2: error: ld returned 1 exit status
```
Not sure if it is problem of dune itself, or the torch package.

**Comment by kit-ty-kate** [comment]
@XVilka the torch issues are unrelated, those packages were already broken before: http://check.ocamllabs.io/log/1559472678-6dfa46c105df9419d2c550c8ccbe44f742586d1f/4.06.1/bad/torch.0.1

**Comment by kit-ty-kate** [comment]
Looks ok, thanks a lot!

**Pull request merged**